rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserParentheseDiscoverer.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.groovy;
+
+import org.codehaus.groovy.ast.expr.MethodCallExpression;
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.internal.StringUtils;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class GroovyParserParentheseDiscoverer {
+    // Matches a code block including leading and trailing parenthesis
+    // Eg: ((((a.invoke~~>("arg")))))<~~ or ~~>(((((( "" as String )))))<~~.toString())
+    private static final Pattern PARENTHESES_GROUP = Pattern.compile(".*?(\\(+[^()]+\\)+).*", Pattern.DOTALL);
+    // Matches when string consist of multiple parenthese blocks
+    // Eg: `((a.invoke())` does not match, `((a.invoke() && b.invoke())` does match
+    private static final Pattern HAS_SUB_PARENTHESIS = Pattern.compile("\\(+[^)]+\\)+[^)]+\\).*", Pattern.DOTALL);
+
+    /**
+     * Calculates the insideParenthesesLevel for method calls, because the compiler does not set the _INSIDE_PARENTHESES_LEVEL flag for method calls.
+     */
+    public static @Nullable Integer getInsideParenthesesLevel(MethodCallExpression node, String source, int cursor) {
+        String sourceFromCursor = source.substring(cursor);
```

Comment:
I'm somewhat surprised to see `source` and `cursor` as separate arguments here when they are merely used to do this substring. Would it make sense to change the method signature here to
```suggestion
    public static @Nullable Integer getInsideParenthesesLevel(MethodCallExpression node, String sourceFromCursor) {
```

Reply:
I had that before, but reverted it back to current state. I think it more clear if the _GroovyParserParentheseDiscoverer_ is a completely separate entity. So no former logic is needed to determine the wrapping. Just pass everything you have (thus the node, the source code of the entire file and the current place you are parsing right now) and let the function discover it all.

Reply:
I wonder if this can be simplified. For Python I had the same basic problem to solve, but I didn't have any "parentheses level" from the AST or similar. Additionally, in Python the parentheses for tuples and generator expressions are sometimes optional, which doesn't make things simpler.

Anyway, rather than counting the "parentheses levels" up front, I added something like a `maybeParenthesized(node, delegate)` method which does something like this:
1. Parse whitespace (and comments) at cursor into `Space`
2. Check if the next character is a `(`
3. If not, reset cursor to before whitespace and return result from `delegate` and we are done
4. Otherwise push a lambda on a parentheses stack, which is a field on the parser visitor
5. Call the delegate
6. Consume whitespace and check if we are at a `)`
7. If yes, pop lambda from parentheses stack and apply it to result, which will wrap it in a `J.Parentheses`

For a language like Java or Python, where we can expect the source files to be reasonably large, I would definitely do something like this, as the substringing and regexp matching won't be free in terms of performance. But for Groovy that would probably not be a big concern, as scripts tend to be small and few.

I will let you think a bit about this and don't want to force you into a different solution.

Reply:
I second Knut's comments and like the approach he describes. I worry that these regexes will be slow and error prone.

Reply:
The one tricky situation I can think of are Java-style type casts, as the first token there is also a `(`. So if the AST node is a Java-style type cast (assuming they can be distinguished from Groovy-style type casts in the AST), the logic could for example pop that last element from the stack and reset the cursor before calling the corresponding visit method (or do it in the method itself). There are also other alternative solutions.

Reply:
I agree, your implementation seems like a better idea @knutwannheden. I would rather first merge this and then do a second round with improvements (or better said, go for the new/better implementation). Currently we don't support nested parenthesis, which just breaks scripts that use them. Having this sub-ideal implementation makes them at least runnable, though with a performance loss. 

Then I can start working on the better implementation.

Reply:
It isn't quite as easy as I made it out to be and I now remember the struggles I had with Python here. The difficulty arises because we don't know at what AST level the parentheses are. Here that parentheses level may come in handy.